lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -897,6 +902,202 @@ matchWitness(TypeChecker &tc,
+
+  // Add the generic signature of the context of the conformance. This includes
+  // the generic parameters from the conforming type as well as any additional
+  // constraints that might occur on the extension that declares the
+  // conformance (i.e., if the conformance is conditional).
+  unsigned depth = 0;
+  if (auto *conformanceSig = conformanceDC->getGenericSignatureOfContext()) {
+    // Use the canonical signature here.
+    conformanceSig = conformanceSig->getCanonicalSignature();
+    allGenericParams.append(conformanceSig->getGenericParams().begin(),
+                            conformanceSig->getGenericParams().end());
+    builder.addGenericSignature(conformanceSig, nullptr);
+    depth = allGenericParams.back()->getDepth() + 1;
+  }
+
+  // FIXME: Think through this
+  //   https://github.com/slavapestov/swift/commit/7f0fa137105915708d81ad373a8231c519a94357
+
+  // Add the generic signature of the requirement, substituting our concrete
+  // type for 'Self'. We don't need the 'Self' requirement or parameter.
+  auto reqDC = req->getInnermostDeclContext();
+  auto reqSig = reqDC->getGenericSignatureOfContext();
+  auto reqEnv = reqDC->getGenericEnvironmentOfContext();
+
+  // First, add the generic parameters from the requirement, not including
+  // the protocol \c Self at the beginning.
+  for (auto genericParam : reqEnv->getGenericParams().slice(1)) {
+    // Create an equivalent generic parameter at the next depth.
+    auto substGenericParam =
+      GenericTypeParamType::get(depth, genericParam->getIndex(), ctx);
+
+    allGenericParams.push_back(substGenericParam);
+    builder.addGenericParameter(substGenericParam);
+
+    // If the generic parameter and its substitution are equivalent, there's
+    // nothing more to do.
+    if (genericParam->isEqual(substGenericParam)) continue;
+
+    // Create a substitution from the requirement's generic parameter to the
+    // generic parameter known to the builder.
+    reqSubs.addSubstitution(genericParam->getCanonicalType(),
+                             substGenericParam);
+    if (auto archetypeType
+          = reqEnv->mapTypeIntoContext(genericParam)->getAs<ArchetypeType>()) {
+      // Add substitutions.
+      SmallVector<ProtocolConformanceRef, 1> conformances;
+      for (auto proto : archetypeType->getConformsTo())
+        conformances.push_back(ProtocolConformanceRef(proto));
+      reqSubs.addConformances(genericParam->getCanonicalType(),
+                               ctx.AllocateCopy(conformances));
+    }
+  }
+
+  // If there were no generic parameters, we're done.
+  if (allGenericParams.empty()) {
+    reqGenericSig = nullptr;
+    return nullptr;
+  }
+
+  // Next, add each of the requirements (mapped from the requirement's
+  // interface types into the abstract type parameters).
+  RequirementSource source(RequirementSource::Explicit, SourceLoc());
+  for (auto &reqReq : reqSig->getRequirements().slice(1)) {
```

Comment:
Isn't the first requirement the witness marker for Self, and the second requirement the conformance requirement for Self : Proto?